[Chee Fong] iP

NCF3535 opened the pull request.

NCF3535 committed:
Ignore .class files

NCF3535 committed:
Rename Duke to Aether and implement basic greeting and exit

NCF3535 committed:
Implement Level-1: Echo functionality in Aether

Aether currently outputs static content and lacks interactivity

Refactor Aether to introduce a command loop that handles user input, echoing commands back and exiting on "exit" as required by Level-1

Transition the codebase to an object-oriented design, separating the logic into the Aether and Display classes to improve code organization and allow for future enhancements

NCF3535 committed:
Add task storage and listing functionality to Aether

Aether previously only echoed user input and exited on "exit," without
storing any user commands

The lack of task storage and listing limited the chatbot's usefulness
as it couldn't track or manage user tasks

Changes made:
* Store user commands as tasks in a fixed-size array, allowing them to
  be listed later
* Introduce a "list" command to display all stored tasks with their
  indices
* Update the program to exit on "bye" instead of "exit" to make it feel
  more conversational and natural for users
* Refactor output handling into the Display class to separate concerns
  and maintain clean code

These changes improve the utility of Aether by enabling task management
and making interactions more intuitive and user-friendly

NCF3535 committed:
Implement marking and unmarking tasks in Aether

Aether currently allows adding and listing tasks, but lacks the
ability to mark tasks as done or revert them to not done.

Marking tasks and umarking as done is crucial for tracking task completion.

Implement functionality to:
*Mark tasks as done, visually indicated by a ✓.
*Unmark tasks, reverting their status to not done.

These changes enhance Aether's task management capabilities,
allowing users to accurately track their tasks.

NCF3535 committed:
Align code with Java coding standards

The code was not fully compliant with the Java coding standards, impacting readability

This commit makes the following adjustments:
*Added inline comments for key code segments to clarify their
  purpose and logic.
*Improved code formatting, including indentation and line wrapping

These changes will help enhance the readability of the code, making it easier to add new features to it in the future

NCF3535 committed:
Add support for Todo, Deadline, and Event tasks

Aether previously only supported generic tasks, limiting its ability
to handle more complex task management needs like deadlines or events.

Changes made:
* Introduced `Todo`, `Deadline`, and `Event` task types, each inheriting
  from `Task`
* Implemented `todo`, `deadline`, and `event` commands for task creation.
* Refactored `addTask`, `listTasks`, and `markTaskStatus` to handle
  task types
* Updated `toString()` methods for custom output of each task type

These changes improve Aether’s task management capabilities, allowing
users to manage different task types with relevant details.

NCF3535 committed:
Add Automated Text UI Testing to Aether

Automated Text UI Testing was introduced using input/output
redirection.

Changes made:
* Implemented error handling with `IF ERRORLEVEL` to display
  "TEST FAILURE" or "Test passed!".
* Added `pause` to prevent the terminal from closing immediately
  after test execution.

These changes improve the testing process by enabling semi-automated
Text UI Testing for Aether, making functionality verification easier.

NCF3535 committed:
Improve code quality and add task count display

Refactored the code to improve readability and consistency.

Changes made:
* Refactored methods to enhance code readability and maintainability.
* Added functionality to display the total number of tasks after each
  task is added.
* Fixed bug where task count was not displayed after adding a task.

These changes improve the code's overall readability while providing
feedback to the user on task count and fixing related issues.

lowjunchen commented on src/main/java/Aether.java:
```diff
@@ -0,0 +1,206 @@
+            String[] eventParts = arguments.split(" /from | /to ");
+            addTask(new Event(eventParts[0], eventParts[1], eventParts[2]));
+            break;
+        default:
```
Good job! You used default in the switch statement

MayFairMI6 replied:
Good job with the indentation around the switch statement

lowjunchen commented on src/main/java/Aether.java:
```diff
@@ -0,0 +1,206 @@
+import java.util.Scanner;
+
+public class Aether {
+    private boolean isExit = false;
```
Good job for making the boolean variable name sound like a boolean

lowjunchen commented on src/main/java/Aether.java:
```diff
@@ -0,0 +1,206 @@
+    private boolean isExit = false;
+    private static final int MAX_TASKS = 100;
+    private Task[] tasks = new Task[MAX_TASKS];
+    private int taskCount = 0;
```
Good job writing variable names as camelcase

lowjunchen commented on src/main/java/Aether.java:
```diff
@@ -0,0 +1,206 @@
+
+        // Trim leading and trailing spaces and split the command into parts
+        command = command.trim();
+        String[] commandParts = command.split(" ", 2);
```
Cannot use magic number here

lowjunchen reviewed (commented):
Good job overall, Chee Fong

MayFairMI6 commented on src/main/java/Aether.java:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    private void handleCommand(String command) {
+        int index = 0;
```
Where are you using this index variable? (I'm assuming it is used to track a specific loop in the function). I would suggest using single-letter variable for tracking iterations, conventionally i, j, k (depends on how nested the loop is) . Otherwise, you have done a good job distinguishing the naming convention in constants (all upper case and underscore) and iterative variables.

MayFairMI6 commented on src/main/java/Aether.java:
```diff
@@ -0,0 +1,206 @@
+public class Aether {
+    private boolean isExit = false;
+    private static final int MAX_TASKS = 100;
+    private Task[] tasks = new Task[MAX_TASKS];
```
Good job mentioning class names such as Aether, Task in PascalCase

MayFairMI6 replied:
plural form of the class name - 'tasks'- enhances readability by giving a clear picture of what is being stored. Keep this up!

MayFairMI6 commented on src/main/java/Aether.java:
```diff
@@ -0,0 +1,206 @@
+    private Task[] tasks = new Task[MAX_TASKS];
+    private int taskCount = 0;
+
+    public static void main(String[] args) {
+        Aether aether = new Aether();
+        aether.run();
+    }
+
+    public void run() {
+        Display.showStartScreen();
+        Scanner scanner = new Scanner(System.in);
+
+        // Main loop to keep the program running until the user exits
+        while (!isExit) {
+            System.out.println("You:");
+            String command = scanner.nextLine();
+            Display.printSeparator();
+            handleCommand(command);
+        }
+    }
```
Consistency in use of braces in code

MayFairMI6 commented on src/main/java/Aether.java:
```diff
@@ -0,0 +1,206 @@
+                System.out.println((i + 1) + "." + tasks[i]);
+            }
+        }
+    }
```
Loops and conditions wrapped appropriately using curly brackets even for single code blacks as per Java coding standard. This provides clarity to the reader and makes the code less bug-prone.

MayFairMI6 reviewed (commented):
Sounds good!
However, minor naming changes might be needed, and some variables should also be made redundant if not necessary. More comments can provide clarity in this scenario. Also, take note that comments should ideally be  in javadoc format ( /**/)

NCF3535 committed:
Add error handling and input validation to Aether

Changes made:
* Added error messages for invalid commands and missing arguments.
* Refactored command handling for better readability.
* Closed Scanner at the end of the program to prevent resource leaks.

These changes improve user experience and make Aether more robust.

NCF3535 committed:
Use proper exceptions to handle errors in Aether

Aether now uses exceptions to handle errors effectively.

Changes:
* Added DukeException class to handle application-specific errors.
* Implemented proper error handling for invalid commands, empty inputs,
  invalid task numbers, and incorrect formats.
* Refactored command logic to throw and catch exceptions as needed.

These changes improve error handling in Aether by using proper
exceptions for better reliability.

NCF3535 committed:
Merge branch 'branch-Level-5'

NCF3535 committed:
Organize classes into suitable packages and update runtest.bat script

Previously, all classes were stored in the same folder, making the
project harder to manage and extend. There was no clear separation of
concerns, and the runtest.bat script did not account for any package
structure, leading to build issues.

Changes made:
* Organized classes into appropriate packages:
  - Moved Aether and DukeException to the 'aether' package.
  - Moved Task, Todo, Deadline, and Event classes to 'aether.task' package.
  - Moved Display class to 'aether.ui' package.
* Updated package declarations in all classes to reflect the new structure.
* Updated runtest.bat:
  - Modified compilation to include 'aether', 'aether.task', and 'aether.ui'.
  - Adjusted execution command to run the 'aether.Aether' main class.
  - Ensured proper comparison of test outputs (ACTUAL.TXT vs EXPECTED.TXT).

These changes improve the organization of the codebase, making it more
maintainable and easier to extend in the future, while ensuring the
test script is aligned with the new package structure.

NCF3535 committed:
Merge branch 'branch-A-Packages'

NCF3535 committed:
Add support for deleting tasks in Aether and use Java Collections

Previously, Aether used a fixed-size array to store tasks, limiting its
ability to handle dynamic task lists and lacked the functionality to delete
tasks.

Changes made:
* Replaced the fixed-size array with ArrayList<Task> for task storage.
* Implemented the delete command to allow users to remove tasks from the
  list.
* Enhanced error handling and input validation for better user experience.
* Standardized error messages for consistency and clarity.

These changes improve Aether's functionality by enabling dynamic task
management using Java Collections and allowing users to delete tasks,
enhancing overall usability.

NCF3535 committed:
Add support for saving and loading tasks in Aether

Previously, Aether did not persist tasks between sessions, limiting its
usefulness.

Changes made:
* Implemented saving tasks to a file whenever the task list changes.
* Implemented loading tasks from the file when Aether starts up.
* Added a Storage class to handle reading from and writing to the file.
* Ensured file paths are relative and OS-independent.
* Handled cases where the data file or directory doesn't exist.

These changes enhance Aether by allowing it to remember tasks across
sessions, improving user experience.

NCF3535 committed:
Merge branch 'branch-Level-6'

NCF3535 committed:
Resolve merge conflicts between Level-6 and Level-7

Changes made:
* Resolve conflicts in Aether.java and Storage.java related to combining
  task deletion with task persistence features.
* Ensure task deletions are reflected in the saved file and handled during
  loading on startup.

This commit integrates task management and persistence functionality.

NCF3535 committed:
Add .gitignore for data file and update MANIFEST.MF

Previously, the data files were not excluded from version control, and
the MANIFEST.MF file was not part of the project.

Changes made:
* Added a .gitignore file to exclude data files from version control.
* Updated the MANIFEST.MF file in src/main/java/META-INF.
* Ensured the .gitignore is configured correctly for the data directory.

These changes ensure that data files remain local and the project
structure includes the necessary manifest file.

yuanners99 commented on src/main/java/aether/Storage.java:
```diff
@@ -0,0 +1,90 @@
+import aether.task.Event;
+
+import java.io.IOException;
+import java.nio.file.*;
```
Try to import explicitly.

yuanners99 commented on src/main/java/aether/Aether.java:
```diff
@@ -0,0 +1,206 @@
+ * Aether class represents the main logic for a simple task manager program.
+ * It allows users to add, delete, mark, unmark tasks, and display tasks.
+ */
+public class Aether {
```
I know it's a work in progress, but remember to add more comments for this class.

yuanners99 reviewed (commented):
Great job! Your code is well-structured and easy to read :)

NCF3535 committed:
Refactor Aether to use more OOP principles

Previously, Aether was not fully utilizing object-oriented principles,
and key functionalities were directly embedded in the main class.

Changes made:
* Extracted the following classes:
    * Ui: Handles interactions with the user.
    * Storage: Manages task loading and saving to the file.
    * Parser: Parses user commands.
    * TaskList: Manages the task list (e.g., adding, deleting,
       marking tasks).
* Implemented Command classes to encapsulate
   task-related operations.
* Refactored the main loop to handle commands dynamically
   and improve scalability.
* Enhanced error handling for file operations and invalid input cases.

These changes make Aether more modular, improving maintainability
and allowing easier future expansions.

NCF3535 committed:
Fix command and storage bugs

* Added missing Event and Deadline cases
* Fixed 'toDataString()' method return return in every task class to
  store in standard format

These changes allow users to correctly add deadlines and events.

NCF3535 committed:
Implement flexible date and time parsing

* Added support for recognizing and parsing date/time formats
* Deadlines are now stored using java.time.LocalDate or LocalDateTime

This enhances the chatbot’s ability to handle and process dates accurately.

NCF3535 committed:
Add Find Command

* Add FindCommand class to support searching tasks by keyword.
* Update TaskList to include findTasks method.
* Integrate find command into the Parser.

These changes allow users to search for tasks by keywords.

NCF3535 committed:
Merge pull request #1 from NCF3535/branch-Level-8

Implement flexible date and time parsing

NCF3535 committed:
Merge pull request #2 from NCF3535/branch-Level-9

Add Find Command

NCF3535 committed:
Add JavaDoc comments to classes and public methods

Previously, Aether lacked comprehensive documentation, making it
challenging to understand

Changes made:
* Added JavaDoc comments to all classes and public methods.

These changes improve Aether by providing clear and thorough
documentation, enhancing readability.

NCF3535 committed:
Merge pull request #3 from NCF3535/branch-A-JavaDoc

Add JavaDoc comments to classes and public methods

NCF3535 committed:
Update README.md

Previously, Aether did not have a user guide, making it difficult
for users to understand how to use the application.

Changes made:
* Added a comprehensive user guide detailing features and usage
  instructions.

These changes improve Aether by providing users with clear
instructions and enhancing usability.

NCF3535 committed:
Correct README update to /docs directory

Previously, the root README.md was updated by mistake, preventing the
GitHub website from reflecting the changes.

Changes made:
* Updated README.md in the /docs directory with the user guide.
* Removed incorrect updates from the root README.md.

These changes ensure that the user guide is properly displayed on the
GitHub website.

NCF3535 committed:
Update README.md for formatting and clarity

Previously, README.md had formatting issues after converting to HTML and lacked key sections.

Changes made:
* Fixed formatting issues to improve readability.
* Added FAQ section to address common questions.
* Added "Editing the Data File" section for user guidance.

These changes enhance README.md by improving clarity and providing
additional information.

NCF3535 committed:
Fix blank todo exception and enhance date parsing

Previously, creating a todo without input did not throw an exception,
causing unexpected behavior.

Changes made:
* Added exception handling for blank todo entries.
* Improved date and time parsing to support additional formats.

These changes ensure that blank todos are properly handled and enhance
flexibility in date and time inputs.

NCF3535 committed:
Update runtest.bat for latest program version

Previously, runtest.bat did not reflect the latest version of the
program, causing discrepancies in test results.

Changes made:
* Updated runtest.bat to test the latest version of the program.
* Adjusted EXPECTED.TXT to match the current program outputs.

These changes ensure that automated tests accurately reflect the
current state of the program.

NCF3535 committed:
Update README.md with supported DATE_TIME formats

Previously, README.md did not specify the supported DATE_TIME
formats nor the behavior when invalid formats are used.

Changes made:
* Listed supported DATE_TIME formats (e.g., YYYY-MM-DD,
  DD/MM/YYYY).
* Explained error handling for unsupported DATE_TIME formats.

These updates ensure users know which DATE_TIME formats are accepted
and understand the application's response to invalid inputs.

NCF3535 committed:
Fix DATE_TIME storage format in toDataString

Previously, DATE_TIME was stored in an incorrect format, causing
unwanted behavior.

Changes made:
* Updated toDataString method to use the desired DATE_TIME format.
* Ensured consistency across all storage operations involving
   DATE_TIME.

These changes ensure that DATE_TIME is stored correctly, preventing
format-related issues.

NCF3535 committed:
Fix README.md formatting for command format notes

Previously, README.md had formatting issues in the notes section about
command formats, making it hard to read.

Changes made:
* Corrected formatting under the command format notes section.
* Improved clarity and readability of command usage examples.

These updates ensure that the command format notes are clearly
presented and easy to understand.
